### ISSUE TITLE
Error when trying to install pyterrier on Windows

### ISSUE THREAD

**Issue by brendasalenave** [terrier-org/pyterrier#308]
I am trying to install pyterrier in my Windows 10 environment; however, it gives me the following error:

```
$ pip install --upgrade git+https://github.com/terrier-org/pyterrier.git#egg=python-terrier
Collecting python-terrier
  Cloning https://github.com/terrier-org/pyterrier.git to c:\users\User\appdata\local\temp\pip-install-i_rtpd7e\python-terrier_4c46d18b35734249b57c84a736906285
  Running command git clone --filter=blob:none --quiet https://github.com/terrier-org/pyterrier.git 'C:\Users\User\AppData\Local\Temp\pip-install-i_rtpd7e\python-terrier_4c46d18b35734249b57c84a736906285'
  Resolved https://github.com/terrier-org/pyterrier.git to commit 943764f32fc04ae0f164f3ec1a399cd28ea8e94b
  Preparing metadata (setup.py) ... done
Requirement already satisfied: numpy in c:\users\User\appdata\local\programs\python\python39\lib\site-packages (from python-terrier) (1.21.5)
Requirement already satisfied: pandas in c:\users\User\appdata\local\programs\python\python39\lib\site-packages (from python-terrier) (1.4.0)
Requirement already satisfied: wget in c:\users\User\appdata\local\programs\python\python39\lib\site-packages (from python-terrier) (3.2)
Requirement already satisfied: tqdm in c:\users\User\appdata\local\programs\python\python39\lib\site-packages (from python-terrier) (4.63.0)
Collecting pyjnius~=1.3.0
  Using cached pyjnius-1.3.0.0.tar.gz (46 kB)
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [16 lines of output]
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "C:\Users\User\AppData\Local\Temp\pip-install-i_rtpd7e\pyjnius_7ac83cbb7bd047ff9e39d01b8c44c751\setup.py", line 95, in <module>
          compile_native_invocation_handler(*get_possible_homes(PLATFORM))
        File "C:\Users\User\AppData\Local\Temp\pip-install-i_rtpd7e\pyjnius_7ac83cbb7bd047ff9e39d01b8c44c751\setup.py", line 85, in compile_native_invocation_handler
          subprocess.check_call([
        File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\subprocess.py", line 368, in check_call
          retcode = call(*popenargs, **kwargs)
        File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\subprocess.py", line 349, in call
          with Popen(*popenargs, **kwargs) as p:
        File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\subprocess.py", line 951, in __init__
          self._execute_child(args, executable, preexec_fn, close_fds,
        File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\subprocess.py", line 1420, in _execute_child
          hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
      PermissionError: [WinError 5] Access is denied
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

My JAVA_HOME environment variable is already defined. I tried to install it by using `pip install python-terrier` as well, and the error is the same. I also tried to run the installation commands by opening the terminal as administrator, which gives me the same error. 

Can anyone help me to solve it?

**Comment by cmacdonald** [comment]
I think you have not got Java installed properly or it's a JRE rather than a JDK. I'd recommend openjdk.

The command that failed was jnius looking for the javac executable.

**Comment by chrissorros** [comment]
I am also getting the same error and I can confirm I have the correct version installed.

**Comment by cmacdonald** [comment]
The problem is in the pyjnius code - perhaps post an issue there and see if anyone has observed it.

If you can help debug, printing the command being executed in https://github.com/kivy/pyjnius/blob/master/setup.py#L78-L92 would be useful to see what its trying to execute.

You can also try setting the JAVA_HOME environment variable manually - perhaps its not finding the Java.

**Comment by cmacdonald** [comment]
Jnius 1.3 isnt available for Python 3.10. We're working on a patch of PyTerrier that works with Jnius 1.4, which should resolve this also. Anyone finding this issue in the meantime should use Python 3.9 or older.

**Comment by brendasalenave** [comment]
For me, the error also occurred when using python 3.9. It is necessary a downgrade to Python 3.8

**Comment by cmacdonald** [comment]
@brendasalenave 
Thanks for your note. The github action testing, e.g https://github.com/terrier-org/pyterrier/actions/runs/2798623261 includes Python 3.9 on Windows. So I expect that to be supported. A bit more about your environment would be useful to see what went wrong there? E.g. could you post the full output when you try to pip install on your Python 3.9.

**Comment by JohnGiorgi** [comment]
Also having trouble installing on MacOS in an environment running python3.9 because of `pyjnius`

```bash
Installing collected packages: pyjnius, ir_measures, ir_datasets, deprecation, statsmodels, sklearn, python-terrier
  Attempting uninstall: pyjnius
    Found existing installation: pyjnius 1.4.2
    Uninstalling pyjnius-1.4.2:
      Successfully uninstalled pyjnius-1.4.2
  Running setup.py install for pyjnius ... error
  error: subprocess-exited-with-error
  
  × Running setup.py install for pyjnius did not run successfully.
  │ exit code: 1
  ╰─> [18 lines of output]
      warning: [options] bootstrap class path not set in conjunction with -source 7
      warning: [options] source value 7 is obsolete and will be removed in a future release
      warning: [options] target value 7 is obsolete and will be removed in a future release
      warning: [options] To suppress warnings about obsolete options, use -Xlint:-options.
      4 warnings
      running install
      /Users/johngiorgi/Library/Caches/pypoetry/virtualenvs/retrieval-exploration-C-nHY9Y1-py3.9/lib/python3.9/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
        warnings.warn(
      running build
      running build_py
      running build_ext
      skipping 'jnius/jnius.c' Cython extension (up-to-date)
      building 'jnius' extension
      jnius/jnius.c:746:10: fatal error: 'jni.h' file not found
      #include "jni.h"
               ^~~~~~~
      1 error generated.
      error: command '/usr/bin/clang' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  Rolling back uninstall of pyjnius
  Moving to /Users/johngiorgi/Library/Caches/pypoetry/virtualenvs/retrieval-exploration-C-nHY9Y1-py3.9/lib/python3.9/site-packages/__pycache__/jnius_config.cpython-39.pyc
   from /private/var/folders/g5/4f6qx6cd5nngsv1qkqv541wr0000gn/T/pip-uninstall-mkl0d83q/jnius_config.cpython-39.pyc
  Moving to /Users/johngiorgi/Library/Caches/pypoetry/virtualenvs/retrieval-exploration-C-nHY9Y1-py3.9/lib/python3.9/site-packages/__pycache__/setup_sdist.cpython-39.pyc
   from /private/var/folders/g5/4f6qx6cd5nngsv1qkqv541wr0000gn/T/pip-uninstall-mkl0d83q/setup_sdist.cpython-39.pyc
  Moving to /Users/johngiorgi/Library/Caches/pypoetry/virtualenvs/retrieval-exploration-C-nHY9Y1-py3.9/lib/python3.9/site-packages/jnius/
   from /Users/johngiorgi/Library/Caches/pypoetry/virtualenvs/retrieval-exploration-C-nHY9Y1-py3.9/lib/python3.9/site-packages/~nius
  Moving to /Users/johngiorgi/Library/Caches/pypoetry/virtualenvs/retrieval-exploration-C-nHY9Y1-py3.9/lib/python3.9/site-packages/jnius_config.py
   from /private/var/folders/g5/4f6qx6cd5nngsv1qkqv541wr0000gn/T/pip-uninstall-r91xw64y/jnius_config.py
  Moving to /Users/johngiorgi/Library/Caches/pypoetry/virtualenvs/retrieval-exploration-C-nHY9Y1-py3.9/lib/python3.9/site-packages/pyjnius-1.4.2.dist-info/
   from /Users/johngiorgi/Library/Caches/pypoetry/virtualenvs/retrieval-exploration-C-nHY9Y1-py3.9/lib/python3.9/site-packages/~yjnius-1.4.2.dist-info
  Moving to /Users/johngiorgi/Library/Caches/pypoetry/virtualenvs/retrieval-exploration-C-nHY9Y1-py3.9/lib/python3.9/site-packages/setup_sdist.py
   from /private/var/folders/g5/4f6qx6cd5nngsv1qkqv541wr0000gn/T/pip-uninstall-r91xw64y/setup_sdist.py
error: legacy-install-failure

× Encountered error while trying to install package.
╰─> pyjnius

note: This is an issue with the package mentioned above, not pip.
hint: See above for output from the failure.
```

**Comment by cmacdonald** [comment]
@JohnGiorgi I think you dont have the appropriate compilers(?) installed, so its cant build pyjnius 1.3. Pyjnius 1.4 has native builds for MacOS on Python 3.9 and 3.10.  Can you verify the branch at https://github.com/terrier-org/pyterrier/pull/249 works for you?

**Comment by JohnGiorgi** [comment]
@cmacdonald Works! Thanks for the amazingly fast response.

**Comment by brendasalenave** [comment]
This is the output that I get after running `pip install python-terrier` command in python3.9 on a windows 10 OS:

```
Collecting python-terrier
  Using cached python-terrier-0.8.1.tar.gz (97 kB)
  Preparing metadata (setup.py) ... done
Requirement already satisfied: numpy in c:\users\brend\appdata\local\programs\python\python39\lib\site-packages (from python-terrier) (1.21.0)
Requirement already satisfied: pandas in c:\users\brend\appdata\local\programs\python\python39\lib\site-packages (from python-terrier) (1.4.0)
Requirement already satisfied: wget in c:\users\brend\appdata\local\programs\python\python39\lib\site-packages (from python-terrier) (3.2)
Requirement already satisfied: tqdm in c:\users\brend\appdata\local\programs\python\python39\lib\site-packages (from python-terrier) (4.63.0)
Collecting pyjnius~=1.3.0
  Using cached pyjnius-1.3.0.0.tar.gz (46 kB)
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [16 lines of output]
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "C:\Users\brend\AppData\Local\Temp\pip-install-7mlgyq80\pyjnius_911da27e62614de19bae58533b618157\setup.py", line 95, in <module>
          compile_native_invocation_handler(*get_possible_homes(PLATFORM))
        File "C:\Users\brend\AppData\Local\Temp\pip-install-7mlgyq80\pyjnius_911da27e62614de19bae58533b618157\setup.py", line 85, in compile_native_invocation_handler
          subprocess.check_call([
        File "C:\Users\brend\AppData\Local\Programs\Python\Python39\lib\subprocess.py", line 368, in check_call
          retcode = call(*popenargs, **kwargs)
        File "C:\Users\brend\AppData\Local\Programs\Python\Python39\lib\subprocess.py", line 349, in call
          with Popen(*popenargs, **kwargs) as p:
        File "C:\Users\brend\AppData\Local\Programs\Python\Python39\lib\subprocess.py", line 951, in __init__
          self._execute_child(args, executable, preexec_fn, close_fds,
        File "C:\Users\brend\AppData\Local\Programs\Python\Python39\lib\subprocess.py", line 1420, in _execute_child
          hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
      PermissionError: [WinError 5] Access is denied
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

**Comment by brendasalenave** [comment]
Removing the `~=1.3.0` from pyjnius in the requirements and than installing pyterrier from source code directly was what worked for me to use it on a Windows OS with Python 3.9.

Thank you all!

**Comment by brendasalenave** [comment]
Removing the `~=1.3.0 `from pyjnius in the requirements and than installing pyterrier from source code directly was what worked for me to use it on a Windows OS with Python 3.9.

Thank you all!

**Comment by cmacdonald** [comment]
Hi everyone, we have merged #322, which addresses support upto Python 3.10. This will be included in the next release of PyTerrier.
